Assets/BossRoom/Scripts/Server/Game/Action/DashAttackAction.cs

```diff
@@ -0,0 +1,98 @@
+using MLAPI;
+using MLAPI.Spawning;
+using UnityEngine;
+using UnityEngine.Assertions;
+
+namespace BossRoom.Server
+{
+    /// <summary>
+    /// Causes the attacker to teleport near a target spot, then perform a melee attack. The client
+    /// visualization moves the character locally beforehand, making the character appear to dash to the 
+    /// destination spot.
+    ///
+    /// After the ExecTime has elapsed, the character is immune to damage until the action ends.
+    ///
+    /// Since the "Range" field means "range when we can teleport to our target", we need another
+    /// field to mean "range of our melee attack after dashing". We'll use the "Radius" field of the
+    /// ActionDescription for that.
+    /// </summary>
+    /// <remarks>
+    /// See MeleeAction for relevant discussion about targeting; we use the same concept here: preferring
+    /// the chosen target, but using whatever is actually within striking distance at time of attack.
+    /// </remarks>
+    public class DashAttackAction : Action
+    {
+        private Vector3 m_TargetSpot;
+
+        public DashAttackAction(ServerCharacter parent, ref ActionRequestData data) : base(parent, ref data)
+        {
+            Assert.IsTrue(Description.Radius > 0, $"ActionDescription for {Description.ActionTypeEnum} needs a Radius assigned!");
+        }
+
+        public override bool Start()
+        {
+            // if we don't have a meaningufl stop-spot, try to find one that's a decent distance away
+            if ((Data.TargetIds == null || Data.TargetIds.Length == 0) && (Data.Position == Vector3.zero || Vector3.Distance(m_Parent.transform.position, Data.Position) < 1))
+            {
+                // it seems like the player doesn't have a useful target. Choose a default one based on their current direction, so that they go somewhere!
+                Data.Position = m_Parent.transform.position + m_Parent.transform.forward * Description.Range;
+            }
+
+            // remember the exact spot we'll stop.
+            m_TargetSpot = ActionUtils.GetTeleportDestination(m_Parent.transform, Data.Position, true, Description.Range, Description.Range);
+
+            // snap to face our destination. This ensures the client visualization faces the right way while "pretending" to dash
+            m_Parent.transform.LookAt(m_TargetSpot);
+
+            // tell clients to visualize this action
+            m_Parent.NetState.RecvDoActionClientRPC(Data);
+
+            return ActionConclusion.Continue;
+        }
+
+        public override bool Update()
+        {
+            return ActionConclusion.Continue;
```

Comment:
This might complicate some things with the need of a secondary RPC, but one case where it's not practical to use the dash is the case when the target moves after choosing it as the destination. An example is targetting an imp who is already chasing you: this action will move towards the initial target spot, not the current imp position so the action ends behind the imp thus dealing no damage. 

Wondering if we can get around this by applying damage while we dash?

Reply:
I think it's acceptable to keep it as is. It'd make it a "surprise attack" kind of dash and make it less OP (it's already pretty OP). 
I wouldn't add too much complexity to our solution just for gameplay reasons.